File: 프로그래머스/1/42889. 실패율/실패율.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;

bool com(pair<double,int>&a, pair<double,int>&b){
    if (a.first != b.first)
        return a.first > b.first;
    return a.second < b.second;
}

vector<int> solution(int N, vector<int> stages) {
    vector <int> user(N+1);
    
    for (int stage: stages){
        user[stage-1]++;
    }
    
    double chalinger = (double)stages.size();
    
    vector <pair<double,int>> fail(N);
    // double tmp;
    
    for (int i = 0; i < N; i++){
        if (!user[i] && !chalinger)
            fail[i] = {0, i+1};
        else
            fail[i] = {user[i] / chalinger, i+1};
        // tmp = user[i] / chalinger;
        // cout << user[i] << "/" << chalinger << "="<< tmp << "\n";
        chalinger -= user[i];
    }
    
    sort(fail.begin(), fail.end(),com);
    vector<int> answer;
    
    for (auto f : fail){
        answer.push_back(f.second);
        // cout << f.first << " " << f.second << endl;
    }
    
    return answer;
}

File: 프로그래머스/1/42889. 실패율/실패율.py
def solution(N, stages):
    state = {i: 0.0 for i in range(1, N + 1)}
    cnt = len(stages)
    
    for s in stages:
        if s <= N:
            state[s] += 1
    
    for i, s in enumerate(state, start=1):
        if state[i] == 0 or cnt == 0:
            continue
        fail = state[i] / cnt  
        cnt -= state[i]
        state[i] = fail
        
    return sorted(state, key = lambda x:state[x], reverse=True)